jni/hidesu.c

```diff
@@ -20,29 +22,266 @@ typedef unsigned short int sa_family_t;
 #include <sys/glibc-syscalls.h>
 #include <asm/unistd.h>
 #include <strings.h>
+#include <dirent.h>
+#include <sys/types.h>
+#include <sys/xattr.h>
+
+char zygote_mnt_ns[256], zygote64_mnt_ns[256], init_mnt_ns[256];
 
 //WARNING: Calling this will change our current namespace
 //We don't care because we don't want to run from here anyway
-int disableSu(int pid) {
-	char *path = NULL;
-	asprintf(&path, "/proc/%d/ns/mnt", pid);
-	int fd = open(path, O_RDONLY);
+int disableSu(unsigned int pid) {
+	char buffer[512];
+	char *p, *p2;
+	struct stat sb;
+	int ret = 0;
+	snprintf(buffer, sizeof(buffer), "/proc/%u/ns/mnt", pid);
+
+	// check that process is running and wait untit it has own mount namespace
+	if(stat(buffer, &sb) == -1) return 1;
+	while(isStillZygoteNs(buffer) != 0) {
+		kill(pid, SIGCONT);
+		usleep(20);
+		kill(pid, SIGSTOP);
+	}
```

Comment:
That could be one option. Another could be using [netlink proc events](http://bewareofgeek.livejournal.com/2945.html). AFAIK both of these methods would still allow race conditions to happen (however with much smaller probability). I guess it would be possible to prevent race conditions for sure by patching zygote process memory (inserting breakpoints), but that would be really fragile and platform specific solution... 
Anyway I guess it would be reasonable to improve process detection/waiting only if we would run into situations when root is detected...